include/psa/crypto_values.h

```diff
@@ -892,6 +892,34 @@
 #define PSA_MAC_TRUNCATED_LENGTH(mac_alg)                               \
     (((mac_alg) & PSA_ALG_MAC_TRUNCATION_MASK) >> PSA_MAC_TRUNCATION_OFFSET)
 
+/* In the encoding of a MAC algorithm, the bit corresponding to
+ * PSA_ALG_MAC_MINIMUM_LENGTH_FLAG encodes the fact that the algorithm is
+ * a usage algorithm, which allows any algorithm corresponding to the same
+ * base class and a tag length greater or equal than the one encoded in
+ * PSA_ALG_MAC_TRUNCATION_MASK. */
+#define PSA_ALG_MAC_MINIMUM_LENGTH_FLAG         ((psa_algorithm_t)0x00008000)
+
+/** Macro to build a MAC minimum-tag-length usage algorithm.
+ *
+ * A mininimum-tag-length MAC usage algorithm contains all MAC algorithms
+ * sharing the same base algorithm, and where the tag length of the specific
+ * algorithm is equal to or larger then the usage's minimum tag length.
+ *
+ * \param mac_alg       A MAC algorithm identifier (value of type
+ *                      #psa_algorithm_t such that #PSA_ALG_IS_MAC(\p alg)
+ *                      is true).
+ * \param tag_length    Desired minimum length of the authentication tag in
+ *                      bytes.
+ *
+ * \return              The corresponding MAC usage algorithm with the
+ *                      specified minimum length.
+ * \return              Unspecified if \p alg is not a supported
+ *                      MAC algorithm or if \p tag_length is not valid
+ *                      for the specified MAC algorithm.
```

Comment:
With the current state of the PR, I've tried to make `PSA_ALG_MAC_WITH_MINIMUM_LENGTH_TAG(mac, 0)` behave the same as `PSA_ALG_AEAD_WITH_MINIMUM_LENGTH_TAG(aead, 0)`, meaning it allows any tag length. Basically, by setting the minimum length to 0 you're saying "anything supported by the encoded algorithm is fine by me", even though the underlying APIs wouldn't allow you to calculate/verify a zero-length MAC or tag.

If you want a different definition, then feel free to propose one. I think "at least 0" = any length makes sense.

Reply:
The current definition for the `mac_length` argument to `PSA_ALG_TRNCATED_MAC()` is:

> Desired length of the truncated MAC in bytes. This must be at most the untruncated length of the MAC and must be at least an implementation-specified minimum. The implementation-specified minimum must not be zero.

And the return value definition states:

> Unspecified if `alg` is not a supported MAC algorithm or if `mac_length` is too small or too large for the specified MAC algorithm.

Is there any reason that the minimum length specified in this policy constructor should use a different validity domain? But note specifically that zero is declared to be invalid by the spec. Rationale: this allows the macro evaluation to result in an untruncated MAC if the application specifies `mac_length` of zero - but the app cannot depend on this behaviour.

Note that for `PSA_ALG_AEAD_WITH_SHORTENED_TAG()`, the domain for `tag_length` is defined differently (and should perhaps be aligned with the truncated MACs?):

> Desired length of the authentication tag in bytes.

The return value definition for the AEAD construction states:

> Unspecified if `alg` is not a supported AEAD algorithm or if `tag_length` is not valid for the specified AEAD algorithm.

Permitted tag lengths are specified explicitly for CCM and GCM, though implementations might not support them all.

In practice, I expect that the macro implementations will just populate the length field (bits[21:16]) with the value provided. This is entirely within the definition of "unspecified".

Reply:
In effect, if we adopt the truncated MAC validity domain for the wildcard algorithm constructor then we get most of the change that Gilles proposes (no reference to "valid for the specified MAC algorithm"). It would explicitly make zero an invalid argument, permitting an unspecified implementation result.

Reply:
Updated. Implemented the same validity domain for mac_length in both truncated MAC and minimum-length MAC.

Reply:
I don't like leaving things unspecified if there isn't a good reason. It's ok for an implementation not to support certain tag lengths, but it should do something sensible. Returning `PSA_ERROR_NOT_SUPPORTED` on an operation is sensible. Accepting a shorter tag than what one would expect on a naive reading of the source code is less sensible and could be a security hole. That may be excused if the tag length is “silly”, like `SIZE_MAX-7`, but less so if it's only moderately silly, like asking for a 10-byte-or-more GCM tag (GCM allows 4, 8, and any value between 12 and 16, but not 10; and 10 is allowed for CCM).

In particular, it makes sense for an application to decide the exact algorithm and the minimum tag length independently.
```
#if HAVE_PMUL_ACCELERATION
base_alg = PSA_ALG_GCM;
#else
base_alg = PSA_ALG_CCM;
#endif
minmum_tag_length = 10; // compromise between message length and strength — a 2^80 attack cost is the minimum we're happy with
policy_alg = WHATEVER_THAT_MACRO_IS_CALLED(base_alg, minimum_tag_length);
```
Making this code invalid is surprising and I don't see a good reason in terms of implementation convenience.

Reply:
@athoelke ?

Reply:
@gilles-peskine-arm - Does the range defined by:

> ``min_length``: This must be at most the untruncated length of the MAC and must be at least an implementation-specified minimum. The implementation-specified minimum must not be zero.
>
> *Return*: Unspecified if ``alg`` is not a supported MAC algorithm or if ``min_length`` is too small or too large for the specified MAC algorithm.

not meet your criteria? - it doesn't allow an implementation to reject 10, if 8 and 12 are valid for the policy.

However, this definition permits a different range for each MAC algorithm, which interferes with variations of the use case you illustrate above.

Perhaps defining a minimum ``min_length`` of ``1`` is better here, than allowing an impdef minimum? - given that this is a wildcard policy and not the operational algorithm. I suggest that we keep the current upper constraint that permits unspecified results if the ``min_length`` is larger than the full length MAC.

Reply:
@gilles-peskine-arm Have you given this some thought, such that we can move this along? Would @athoelke 's suggestion of setting the minimum length for which the return value is specified to 1 (or maybe to `the implementation-specified minimum`?) for wildcard policies cover your worry?

Reply:
Making the minimum implementation-specified solves this particular case, assuming the implementation does support short tags. It doesn't work if the implementation requires a minimum of 10 for all AEAD tags, and so 10 is less than the supported minimum for GCM. I can't think of a good reason to make the minimum _permitted in policies_ implementation-specific, in a separate way from the minimum that the implementation permits for actual use.

In general, I prefer to avoid implementation-specified behavior, other than allowing implementations to omit support for a feature altogether. (Here, by “feature”, I include things like specific key sizes which an implementation might omit as a resource limit (maximum key size), a security policy (minimum key size) or an implementation simplification (e.g. AES-192 or odd-size RSA exponent).) If an implementation doesn't support shorter tag lengths, that's fine by me. But I can't think of a good reason for a separate implementation-specific behavior regarding policies.

I'm fine with allowing 1 as a minimum, as in, if an implementation supports `PSA_ALG_WHATEVER_IT_S_CALLED(alg, length)` then it must support e.g. `PSA_ALG_WHATEVER_IT_S_CALLED(alg, 1)` to mean “alg with any supported tag length”.

Reply:
I agree. I would suggest wording that is something along the lines of:

> ``min_length``: This must be at most the untruncated length of the MAC and must be at least ``1``.
>
> **Return:** Unspecified if ``alg`` is not a supported MAC algorithm or if ``min_length`` is less than ``1`` or too large for the specified MAC algorithm.

It might be worth a comment in the description that specifying ``min_length = 1`` will construct a policy that permits all supported MAC/tag lengths.

Reply:
Done.